flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSourceSinkFactoryTestBase.java

```diff
@@ -177,13 +157,112 @@ public void testTableSource() {
+		fieldMapping.put(TIME, TIME);
+
+		final Map<KafkaTopicPartition, Long> specificOffsets = new HashMap<>();
+		specificOffsets.put(new KafkaTopicPartition(TOPIC, PARTITION_0), OFFSET_0);
+		specificOffsets.put(new KafkaTopicPartition(TOPIC, PARTITION_1), OFFSET_1);
+
+		final TestDeserializationSchema deserializationSchema = new TestDeserializationSchema(
+				TableSchema.builder()
+						.field(NAME, DataTypes.STRING())
+						.field(COUNT, DataTypes.DECIMAL(10, 3))
+						.field(TIME, DataTypes.TIMESTAMP(3))
+						.build()
+						.toRowType()
+		);
+
+		final KafkaTableSourceBase expected = getExpectedKafkaTableSource(
+				schema,
+				Optional.of(PROC_TIME),
+				rowtimeAttributeDescriptors,
+				fieldMapping,
+				TOPIC,
+				KAFKA_PROPERTIES,
+				deserializationSchema,
+				StartupMode.SPECIFIC_OFFSETS,
+				specificOffsets);
+
+		TableSourceValidation.validateTableSource(expected);
+
+		// construct table source using descriptors and table source factory
+		final Map<String, String> legacyPropertiesMap = new HashMap<>();
+		legacyPropertiesMap.putAll(createKafkaSourceProperties());
+
+		// use legacy properties
+		legacyPropertiesMap.remove("connector.specific-offsets");
+		legacyPropertiesMap.remove("connector.properties.zookeeper.connect");
+		legacyPropertiesMap.remove("connector.properties.bootstrap.servers");
+		legacyPropertiesMap.remove("connector.properties.group.id");
+
+		legacyPropertiesMap.put("connector.specific-offsets.0.partition", "0");
+		legacyPropertiesMap.put("connector.specific-offsets.0.offset", "100");
+		legacyPropertiesMap.put("connector.specific-offsets.1.partition", "1");
+		legacyPropertiesMap.put("connector.specific-offsets.1.offset", "123");
+		legacyPropertiesMap.put("connector.properties.0.key", "zookeeper.connect");
+		legacyPropertiesMap.put("connector.properties.0.value", "dummy");
+		legacyPropertiesMap.put("connector.properties.1.key", "bootstrap.servers");
+		legacyPropertiesMap.put("connector.properties.1.value", "dummy");
+		legacyPropertiesMap.put("connector.properties.2.key", "group.id");
+		legacyPropertiesMap.put("connector.properties.2.value", "dummy");
+
+		final TableSource<?> actualSource = TableFactoryService.find(StreamTableSourceFactory.class, legacyPropertiesMap)
+				.createStreamTableSource(legacyPropertiesMap);
+
+		assertEquals(expected, actualSource);
+
+		// test Kafka consumer
+		final KafkaTableSourceBase actualKafkaSource = (KafkaTableSourceBase) actualSource;
+		final StreamExecutionEnvironmentMock mock = new StreamExecutionEnvironmentMock();
+		actualKafkaSource.getDataStream(mock);
+		assertTrue(getExpectedFlinkKafkaConsumer().isAssignableFrom(mock.sourceFunction.getClass()));
+	}
+
+	protected Map<String, String> createKafkaSourceProperties() {
+		return new TestTableDescriptor(
+				new Kafka()
+						.version(getKafkaVersion())
+						.topic(TOPIC)
+						.properties(KAFKA_PROPERTIES)
+						.sinkPartitionerRoundRobin() // test if accepted although not needed
```

Comment:
one tab?

Reply:
It's default from flink project code style, I'll try to use one tab.